### PR TITLE
sql: use an implicit txn during the extended protocol

### LINKED ISSUE #71665: sql/pgwire: Sync command should destroy named portals when outside of explicit txn
**Describe the problem**

According to https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
> If successfully created, a named portal object lasts till the end of the current transaction

In other words, after `Sync` is issued and there is no explicit transaction, the portal that was bound using `Bind` is destroyed. However, CRDB doesn't implement this behavior.

Indeed, if you try this test against Postgres, it results in portal "p8" does not exist

**To Reproduce**

The following pgwire test matches the Postgres behavior
```
send
Parse {"Name": "s8", "Query": "SELECT oid, relname FROM pg_class WHERE oid = $1::regclass;"}
Bind {"DestinationPortal": "p8", "PreparedStatement": "s8", "ResultFormatCodes": [0], "Parameters": [{"text":"t"}]}
Sync
----

until
ReadyForQuery
----
{"Type":"ParseComplete"}
{"Type":"BindComplete"}
{"Type":"ReadyForQuery","TxStatus":"I"}

send
Execute {"Portal": "p8"}
Sync
----

until
ErrorResponse
ReadyForQuery
----
{"Type":"ErrorResponse","Code":"34000"}
{"Type":"ReadyForQuery","TxStatus":"I"}
```

Running against CRDB with `make test PKG=./pkg/sql/pgwire TESTS='TestPGTest` fails

**Expected behavior**

The above test should pass

**Additional context**
I wonder if this is related to the OOM we saw in https://github.com/cockroachdb/cockroach/issues/69317

### PR DESCRIPTION
**Pull request by rafiss**
fixes https://github.com/cockroachdb/cockroach/issues/71665

In the Postgres wire protocol, entering the extended protocol also
starts an implicit transaction. This transaction is committed when the
Sync message is received.

Previously, we were starting 3 separate "implicit" transactions for the
Prepare, Bind, and ExecPortal commands. (Implicit is in scare quotes
because they were created entirely ad-hoc.)

Now, the same implicit txn is used for the duration of the extended
protocol.

If BEGIN is executed as a prepared statement, then the implicit
trasnaction is upgraded to an explicit one. This behavior matches
Postgres.

Release note (bug fix): Fixed a bug where the different stages of
preparing, binding, and executing a prepared statement would use
different implicit transactions. Now these stages all share the same
implicit transaction.

### CODE REVIEW

**Comment by rafiss** [comment]
Here's a test that fails before this PR, which I would have expected this PR to fix. But it still fails.

```
func TestResolveTypesAOST(t *testing.T) {
	srv, db, _ := serverutils.StartServer(t, base.TestServerArgs{Insecure: true})
	defer srv.Stopper().Stop(context.Background())
	defer db.Close()

	time.Sleep(2 * time.Second)

	_, err := db.Exec("CREATE TYPE typ AS ENUM('hi', 'hello')")
	require.NoError(t, err)

	var s string

	err = db.QueryRow("SELECT 'hi'::typ FROM generate_series(1,1) AS OF SYSTEM TIME '-1s'").Scan(&s)
	// Fails as expected
	require.EqualError(t, err, "pq: type \"typ\" does not exist")

	err = db.QueryRow("SELECT $1::typ FROM generate_series(1,1) AS OF SYSTEM TIME '-1s'", "hi").Scan(&s)
	// No error happens even though it should
	require.NoError(t, err)

	err = db.QueryRow("SELECT $1::typ FROM generate_series(1,1) AS OF SYSTEM TIME '-1s'", "hi").Scan(&s)
	// The error happens here though.
	require.EqualError(t, err, "pq: type \"typ\" does not exist")

	_, err = db.Exec("CREATE TABLE tab(a TEXT)")
	require.NoError(t, err)

	err = db.QueryRow("SELECT * FROM tab AS OF SYSTEM TIME '-1s'").Scan(&s)
	// No error happens even though it should
	require.NoError(t, err)

	err = db.QueryRow("SELECT * FROM tab AS OF SYSTEM TIME '-1s'").Scan(&s)
	// The error happens here though.
	require.EqualError(t, err, "pq: type \"typ\" does not exist")
}
```

File: pkg/sql/pgwire/testdata/pgtest/pgjdbc [hunk 1]
```diff
@@ -72,3 +72,73 @@ ReadyForQuery
...
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+
+# Test that a simple query in the middle of the extended protocol
+# causes the earlier statement to commit.
+send
+Parse {"Name": "S_4", "Query": "INSERT INTO t VALUES(1)"}
+Bind {"PreparedStatement": "S_4"}
+Execute
+Query {"String": "SELECT 1"}
```
Comment (otan) [inline pkg/sql/pgwire/testdata/pgtest/pgjdbc:113]: can you add a comment that the prepared statement is re-executed and we assume a constraint error from that? that confused me for a second.

**Comment by otan** [comment]
```

	err = db.QueryRow("SELECT * FROM tab AS OF SYSTEM TIME '-1s'").Scan(&s)
	// No error happens even though it should
	require.NoError(t, err)

	err = db.QueryRow("SELECT * FROM tab AS OF SYSTEM TIME '-1s'").Scan(&s)
	// The error happens here though.
	require.EqualError(t, err, "pq: type \"typ\" does not exist") 
```

why isn't the error "tab does not exist?"

**Review by otan — COMMENTED** [review]
took a look, changes makes sense to me. will caveat i'm not 100% on the extended protocol. i trust ORM tests to break if something is off ;)

the error above is suspicious to me though.

File: pkg/sql/pgwire/testdata/pgtest/pgjdbc (hunk 1 above)
Comment (rafiss) [reply pkg/sql/pgwire/testdata/pgtest/pgjdbc:113]: yeah for sure

**Comment by rafiss** [comment]
> why isn't the error "tab does not exist?" 

oops my bad -- copy paste error. also, it seems that one of the other fixes I made to clean up the PR made this test (with typos fixed) pass now, so we're all good!

**Comment by rafiss** [comment]
> i trust ORM tests to break if something is off ;)

I'll try running the ORM tests before merging this

**Review by otan — APPROVED** [review]

**Comment by rafiss** [comment]
tftr!

bors r=otan

**Pull request merged**